### ISSUE TITLE
Document/Verify minimum supported CMake version per build OS

### ISSUE THREAD

**Issue by jkoritzinsky** [dotnet/coreclr#23868]
We require a different version of CMake on each OS platform that we build on since we use the version provided by the package repository on the OS for non-Windows platforms. However, we have no documentation about what versions these are. As a result, it's very difficult to try to clean up our CMake scripts since there's no good way to know what version we need to support until a CI run fails.

cc: @janvorli @jashook

**Comment by janvorli** [comment]
Our requirement is cmake 2.8.12, as recorded in the root CMakeLists.txt:
https://github.com/dotnet/coreclr/blob/f63bb80464e5432fc3250375bc38830d82c531ce/CMakeLists.txt#L2

Our CMake scripts should work on cmake of that version or newer 

The only exception is Windows, but that's not because of the scripts, but because of the version of VS projects / solutions we want it to generate. A CMake version that supports the target version of Visual Studio is required.

**Comment by jkoritzinsky** [comment]
Are there any OS versions that we're still building on that bundle 2.8.12 instead of 3.0 or newer? The last I know of is Ubuntu 14.04 which is EOL, but I don't know if I'm missing a platform. We have a few places in our build scripts that have to branch based on behavior changes in 3.0. Additionally, there are quite a few places throughout our CMake scripts that could be cleaned up if we had access to 3.0 across the board.

**Comment by janvorli** [comment]
> Are there any OS versions that we're still building on that bundle 2.8.12 instead of 3.0 or newer?

Yes, we build arm32 official builds on Ubuntu 14.04 to make it as universal as possible. Similar to why we build for x64 on CentOS 7. On both systems, the 2.8.12 is the highest available version.

**Comment by janvorli** [comment]
I wonder, what do you have on mind as places that could be cleaned up if we moved to 3.0+?

**Comment by jkoritzinsky** [comment]
The one I ran into yesterday was refactoring some of the settings we have in `windows-compiler-override.txt` and family to use new CMake properties like `CMAKE_CXX_STANDARD` or things like `add_compile_options` or the new target-focused commands like `target_include_directories`, `target_compile_definitions`, `target_compile_options` instead of the global commands that affect all targets created afterwards.

There's also the few places that we have that branch on CMake version.

There's also the fact that CMake documentation pre-3.0 is in a different system and harder to track down if you haven't found it previously, so it's difficult to even check if a feature was new in 3.0 or available in 2.8.12. (I finally found it and bookmarked it a few days ago).

I don't think this is enough value proposition to move us up to a new CMake version, but I felt that if we are no longer tied to 2.8.12, we should do the CMake cleanup.

**Comment by jkoritzinsky** [comment]
We should revisit this when Ubuntu 14.04 LTS, RHEL 6, and CentOS 6 hit EOL and .NET Core is no longer supporting them.

Ubuntu 14.04 will at the end of this month and 3.0 isn't going to support it. RHEL 6 and CentOS 6 will hit EOL on November 30, 2020, so we've got a while until they're not supported.

**Comment by janvorli** [comment]
> RHEL 6 and CentOS 6

I was talking about CentOS 7, not 6. 

If we really wanted to move to 3.0+, we could compile a new cmake version ourselves when building the docker images that we use for building stuff. In fact, I have just realized that the arm32 is being cross-built and so we are not using the cmake 2.8.12 anymore. We use a newer version present on the machine we use for cross-building. So the x64 builds on CentOS 7 are the only ones using the 2.8.12.

Btw, I am not sure where we would benefit from new target-focused commands like target_include_directories, target_compile_definitions, target_compile_options instead of the global commands that affect all targets created afterwards. In fact, I consider the hierarchical inheritance of options, definitions and include directories one of the best features of cmake and I haven't hit a case in the past where it would stand in my way. Do you have specific places on your mind where you feel it would be better to use the per-target ones?

**Comment by jkoritzinsky** [comment]
If we're talking CentOS 7, then I think they have a `cmake3` package that installs a 3.X version of CMake under the name `cmake3` that we could use or set up as an alternative symlink for `cmake` and make it default in our images.

The first ideas I have would be the places we have across the repo where we have two subdirectories that only contain CMakeLists.txt files. We could combine those files and use the target-focused commands to only add definitions to specific targets. `mscorrc`'s full and small directories are the first one I found in a quick glance. Our 4 different targets for `utilcode` are another possible example. As well as `unwinder/dac`.

Basically use them in combination with the non-target-focused commands to coalesce our CMake scripts a bit.

**Comment by jkoritzinsky** [comment]
It looks like the CentOS images already install the `cmake3` package to build llvm and clang if they're the ones at https://github.com/dotnet/dotnet-buildtools-prereqs-docker

**Comment by jkoritzinsky** [comment]
I did a little test of the different CMake versions on each of our CI build OS's. Here's what I've found:

```
AzDO:

Windows x64: 3.11.1
Windows x86: 3.11.1
Linux arm: 2.8.12.2
Linux arm64: 3.5.1
Linux_musl x64: 3.8.1
Linux x64: 3.6.3
OSX x64: 3.14.1
Windows arm: 3.11.1
Windows arm64: 3.11.1
Linux_rhel6 x64: 3.6.1 (cmake3)

Jenkins:
Ubuntu ARM crossgen_comparison jobs: 2.8.12.2
Ubuntu x64 CoreFX jobs: 2.8.12.2
Windows x64 CoreFX jobs: 3.9.3
```

So it looks like if we update the Linux arm32 Docker image to be based on Ubuntu 16.04 (which is what the ARM64 image is based on), then our Azure DevOps builds will all be at a minimum CMake version of 3.5.1. Since Ubuntu 14.04 is hitting EOL at the end of April and .NET Core 3.0 won't support Ubuntu 14.04 anyway, we should do this update at some point.

The Jenkins Linux images are still on 2.8.12.2, so my thought would be to just wait for Jenkins to get shut off before we update our CMake scripts to use 3.0 and newer features.
